### DISCUSSION TITLE
Is it possible to access additional EXIF tags

### DISCUSSION THREAD

**Discussion by NickAyesha** [python-pillow/Pillow#7486] [Q&A] [answered]
Hello Everyone!
Can someone have a look and give some assistance here? 

I am working with image files' Exif data, rather than the image itself. Things were going pretty good until I reached a roadblock after having realized that the population of EXIF tags available by default is about 46.

There are some important tags that I need for my project, but. they aren't accessible:

For example:
Tag ID | Tag Name
0x9c9e | XPKeywords
0x9c9f | XPSubject
0xa436 | Title

Is there a way to extend the tags that  _getexif() returns? If you can suggest what direction I need to take in order to access the information I need, then I would really appreciate it.

Thanks

**Comment by radarhere** [comment]
Rather than using `_getexif()`, I suggest you use the newer [`getexif()`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.getexif). This will return an instance of the [`Exif`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Exif) class. Then, `get_ifd()` should let you access the tags in the EXIF IFD that you're after.

```python
with Image.open("Tests/images/flower.jpg") as im:
    exif = im.getexif()
    print(exif.get_ifd(ExifTags.IFD.Exif))
```

**Comment by NickAyesha** [comment]
Hi Radar -I appreciate your quick response.
I tried that method, and while it returns tags, it gives an even smaller subset of tags (32 to be exact).

IFD is defined as

**class IFD(IntEnum)**:
    Exif = 34665
    GPSInfo = 34853
    Makernote = 37500
    Interop = 40965
    IFD1 = -1

 Do you know what the value 34665 stems from, and could that be the key to accessing additional tags? I'm completely new to this project and trying to solve what I initially thought to be a simple problem.

**Reply by radarhere** [reply]
If you look at https://exiftool.org/TagNames/EXIF.html, you will see a list of EXIF tags. One of the columns is 'Group'. EXIF tags are not structured 1 dimensionally, but have groups.

As you are figuring out, 34665 in hexadecimal is 0x8769. At https://exiftool.org/TagNames/EXIF.html, you will see '0x8769' listed as the 'ExifOffset'. This is the way to access the Exif IFD.

**Reply by radarhere** [reply]
I thought you were specifically after the XP tags, but if you would like more, you can try `getexif()` by itself, or other IFDs.
```python
from PIL import Image, ExifTags
with Image.open("Tests/images/flower.jpg") as im:
    exif = im.getexif()
    print(exif)
    print(exif.get_ifd(ExifTags.IFD.Exif))
    print(exif.get_ifd(ExifTags.IFD.GPSInfo))
    print(exif.get_ifd(ExifTags.IFD.Interop))
    print(exif.get_ifd(ExifTags.IFD.IFD1))
```

**Reply by NickAyesha** [reply]
This is exactly the hint I needed. Thank you!

**Reply by NickAyesha** [reply]
Most urgent for me right now are the Title and Keywords, and i think those are sub-groups of EXIF.

**Reply by radarhere** [reply]
https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml lists them as under 'Exif Private IFD', but https://exiftool.org/TagNames/EXIF.html lists them as IFD0. IFD0 is the basic values from `getexif()`.

**Reply by NickAyesha** [reply]
The tags are probably be under 'Exif Private IFD' as you mentioned because I checked IFD0 and didn't find them. 
They may be under the IPTC, specifically.

I will need to find a way to identify the sub-IFDs (groups) and traverse those nodes. Do you know if that is possible with Pillow's latest build?

Code | Name | Short description | Detail
-- | -- | -- | --
Dec | Hex | -- | --
32932 | 80A4 | Wang   Annotation | Annotation data, as used in 'Imaging for Windows'.
33445 | 82A5 | MD   FileTag | Specifies the pixel data format encoding in the Molecular   Dynamics GEL file format.
33446 | 82A6 | MD   ScalePixel | Specifies a scale factor in the Molecular Dynamics GEL file   format.
33447 | 82A7 | MD   ColorTable | Used to specify the conversion from 16bit to 8bit in the   Molecular Dynamics GEL file format.
33448 | 82A8 | MD   LabName | Name of the lab that scanned this file, as used in the   Molecular Dynamics GEL file format.
33449 | 82A9 | MD   SampleInfo | Information about the sample, as used in the Molecular   Dynamics GEL file format.
33450 | 82AA | MD   PrepDate | Date the sample was prepared, as used in the Molecular   Dynamics GEL file format.
33451 | 82AB | MD   PrepTime | Time the sample was prepared, as used in the Molecular   Dynamics GEL file format.
33452 | 82AC | MD   FileUnits | Units for data in this file, as used in the Molecular Dynamics   GEL file format.
33550 | 830E | ModelPixelScaleTag | Used in interchangeable GeoTIFF files.
33723 | 83BB | IPTC | IPTC (International Press Telecommunications Council)   metadata.
33918 | 847E | INGR   Packet Data Tag | Intergraph Application specific storage.
33919 | 847F | INGR   Flag Registers | Intergraph Application specific flags.
33920 | 8480 | IrasB   Transformation Matrix | Originally part of Intergraph's GeoTIFF tags, but likely   understood by IrasB only.
33922 | 8482 | ModelTiepointTag | Originally part of Intergraph's GeoTIFF tags, but now used in   interchangeable GeoTIFF files.
34264 | 85D8 | ModelTransformationTag | Used in interchangeable GeoTIFF files.
34377 | 8649 | Photoshop | Collection of Photoshop 'Image Resource Blocks'.
34665 | 8769 | Exif   IFD | A pointer to the Exif IFD.
34675 | 8773 | ICC   Profile | ICC profile data.
34735 | 87AF | GeoKeyDirectoryTag | Used in interchangeable GeoTIFF files.
34736 | 87B0 | GeoDoubleParamsTag | Used in interchangeable GeoTIFF files.
34737 | 87B1 | GeoAsciiParamsTag | Used in interchangeable GeoTIFF files.
34853 | 8825 | GPS IFD | A pointer to the Exif-related GPS Info IFD.
34908 | 885C | HylaFAX   FaxRecvParams | Used by HylaFAX.
34909 | 885D | HylaFAX   FaxSubAddress | Used by HylaFAX.
34910 | 885E | HylaFAX   FaxRecvTime | Used by HylaFAX.
37724 | 935C | ImageSourceData | Used by Adobe Photoshop.
40965 | A005 | Interoperability   IFD | A pointer to the Exif-related Interoperability IFD.
42112 | A480 | GDAL_METADATA | Used by the GDAL library, holds an XML list of name=value   'metadata' values about the image as a whole, and about specific samples.
42113 | A481 | GDAL_NODATA | Used by the GDAL library, contains an ASCII encoded nodata or   background pixel value.
50215 | C427 | Oce   Scanjob Description | Used in the Oce scanning process.
50216 | C428 | Oce   Application Selector | Used in the Oce scanning process.
50217 | C429 | Oce   Identification Number | Used in the Oce scanning process.
50218 | C42A | Oce   ImageLogic Characteristics | Used in the Oce scanning process.
50706 | C612 | DNGVersion | Used in IFD 0 of DNG files.
50707 | C613 | DNGBackwardVersion | Used in IFD 0 of DNG files.
50708 | C614 | UniqueCameraModel | Used in IFD 0 of DNG files.
50709 | C615 | LocalizedCameraModel | Used in IFD 0 of DNG files.
50710 | C616 | CFAPlaneColor | Used in Raw IFD of DNG files.
50711 | C617 | CFALayout | Used in Raw IFD of DNG files.
50712 | C618 | LinearizationTable | Used in Raw IFD of DNG files.
50713 | C619 | BlackLevelRepeatDim | Used in Raw IFD of DNG files.
50714 | C61A | BlackLevel | Used in Raw IFD of DNG files.
50715 | C61B | BlackLevelDeltaH | Used in Raw IFD of DNG files.
50716 | C61C | BlackLevelDeltaV | Used in Raw IFD of DNG files.
50717 | C61D | WhiteLevel | Used in Raw IFD of DNG files.
50718 | C61E | DefaultScale | Used in Raw IFD of DNG files.
50719 | C61F | DefaultCropOrigin | Used in Raw IFD of DNG files.
50720 | C620 | DefaultCropSize | Used in Raw IFD of DNG files.
50721 | C621 | ColorMatrix1 | Used in IFD 0 of DNG files.
50722 | C622 | ColorMatrix2 | Used in IFD 0 of DNG files.
50723 | C623 | CameraCalibration1 | Used in IFD 0 of DNG files.
50724 | C624 | CameraCalibration2 | Used in IFD 0 of DNG files.
50725 | C625 | ReductionMatrix1 | Used in IFD 0 of DNG files.
50726 | C626 | ReductionMatrix2 | Used in IFD 0 of DNG files.
50727 | C627 | AnalogBalance | Used in IFD 0 of DNG files.
50728 | C628 | AsShotNeutral | Used in IFD 0 of DNG files.
50729 | C629 | AsShotWhiteXY | Used in IFD 0 of DNG files.
50730 | C62A | BaselineExposure | Used in IFD 0 of DNG files.
50731 | C62B | BaselineNoise | Used in IFD 0 of DNG files.
50732 | C62C | BaselineSharpness | Used in IFD 0 of DNG files.
50733 | C62D | BayerGreenSplit | Used in Raw IFD of DNG files.
50734 | C62E | LinearResponseLimit | Used in IFD 0 of DNG files.
50735 | C62F | CameraSerialNumber | Used in IFD 0 of DNG files.
50736 | C630 | LensInfo | Used in IFD 0 of DNG files.
50737 | C631 | ChromaBlurRadius | Used in Raw IFD of DNG files.
50738 | C632 | AntiAliasStrength | Used in Raw IFD of DNG files.
50739 | C633 | ShadowScale | Used in IFD 0 of DNG files.
50740 | C634 | DNGPrivateData | Used in IFD 0 of DNG files.
50741 | C635 | MakerNoteSafety | Used in IFD 0 of DNG files.
50778 | C65A | CalibrationIlluminant1 | Used in IFD 0 of DNG files.
50779 | C65B | CalibrationIlluminant2 | Used in IFD 0 of DNG files.
50780 | C65C | BestQualityScale | Used in Raw IFD of DNG files.
50781 | C65D | RawDataUniqueID | Used in IFD 0 of DNG files.
50784 | C660 | Alias   Layer Metadata | Alias Sketchbook Pro layer usage description.
50827 | C68B | OriginalRawFileName | Used in IFD 0 of DNG files.
50828 | C68C | OriginalRawFileData | Used in IFD 0 of DNG files.
50829 | C68D | ActiveArea | Used in Raw IFD of DNG files.
50830 | C68E | MaskedAreas | Used in Raw IFD of DNG files.
50831 | C68F | AsShotICCProfile | Used in IFD 0 of DNG files.
50832 | C690 | AsShotPreProfileMatrix | Used in IFD 0 of DNG files.
50833 | C691 | CurrentICCProfile | Used in IFD 0 of DNG files.
50834 | C692 | CurrentPreProfileMatrix | Used in IFD 0 of DNG files.
50879 | C6BF | ColorimetricReference | Used in IFD 0 of DNG files.
50908 | C6DC | TIFF_RSID | This private tag is used in a GEOTIFF standard by DGIWG.
50909 | C6DD | GEO_METADATA | This private tag is used in a GEOTIFF standard by DGIWG.
50931 | C6F3 | CameraCalibrationSignature | Used in IFD 0 of DNG files.
50932 | C6F4 | ProfileCalibrationSignature | Used in IFD 0 or Camera Profile IFD of DNG files.
50933 | C6F5 | ExtraCameraProfiles | Used in IFD 0 of DNG files.
50934 | C6F6 | AsShotProfileName | Used in IFD 0 of DNG files.
50935 | C6F7 | NoiseReductionApplied | Used in Raw IFD or Enhanced IFD of DNG files.
50936 | C6F8 | ProfileName | Used in IFD 0 or Camera Profile IFD of DNG files.
50937 | C6F9 | ProfileHueSatMapDims | Used in IFD 0 or Camera Profile IFD of DNG files.
50938 | C6FA | ProfileHueSatMapData1 | Used in IFD 0 or Camera Profile IFD of DNG files.
50939 | C6FB | ProfileHueSatMapData2 | Used in IFD 0 or Camera Profile IFD of DNG files.
50940 | C6FC | ProfileToneCurve | Used in IFD 0 or Camera Profile IFD of DNG files.
50941 | C6FD | ProfileEmbedPolicy | Used in IFD 0 or Camera Profile IFD of DNG files.
50942 | C6FE | ProfileCopyright | Used in IFD 0 or Camera Profile IFD of DNG files.
50964 | C714 | ForwardMatrix1 | Used in IFD 0 or Camera Profile IFD of DNG files.
50965 | C715 | ForwardMatrix2 | Used in IFD 0 or Camera Profile IFD of DNG files.
50966 | C716 | PreviewApplicationName | Used in Preview IFD of DNG files.
50967 | C717 | PreviewApplicationVersion | Used in Preview IFD of DNG files.
50968 | C718 | PreviewSettingsName | Used in Preview IFD of DNG files.
50969 | C719 | PreviewSettingsDigest | Used in Preview IFD of DNG files.
50970 | C71A | PreviewColorSpace | Used in Preview IFD of DNG files.
50971 | C71B | PreviewDateTime | Used in Preview IFD of DNG files.
50972 | C71C | RawImageDigest | Used in IFD 0 of DNG files.
50973 | C71D | OriginalRawFileDigest | Used in IFD 0 of DNG files.
50974 | C71E | SubTileBlockSize | Used in Raw IFD of DNG files.
50975 | C71F | RowInterleaveFactor | Used in Raw IFD of DNG files.
50981 | C725 | ProfileLookTableDims | Used in IFD 0 or Camera Profile IFD of DNG files.
50982 | C726 | ProfileLookTableData | Used in IFD 0 or Camera Profile IFD of DNG files.
51008 | C740 | OpcodeList1 | Used in Raw IFD of DNG files.
51009 | C741 | OpcodeList2 | Used in Raw IFD of DNG files.
51022 | C74E | OpcodeList3 | Used in Raw IFD of DNG files.
51041 | C761 | NoiseProfile | Used in Raw IFD or Enhanced IFD of DNG files.
51089 | C791 | OriginalDefaultFinalSize | Used in IFD 0 of DNG files.
51090 | C792 | OriginalBestQualityFinalSize | Used in IFD 0 of DNG files.
51091 | C793 | OriginalDefaultCropSize | Used in IFD 0 of DNG files.
51107 | C7A3 | ProfileHueSatMapEncoding | Used in IFD 0 or Camera Profile IFD of DNG files.
51108 | C7A4 | ProfileLookTableEncoding | Used in IFD 0 or Camera Profile IFD of DNG files.
51109 | C7A5 | BaselineExposureOffset | Used in IFD 0 or Camera Profile IFD of DNG files.
51110 | C7A6 | DefaultBlackRender | Used in IFD 0 or Camera Profile IFD of DNG files.
51111 | C7A7 | NewRawImageDigest | Used in IFD 0 of DNG files.
51112 | C7A8 | RawToPreviewGain | Used in Preview IFD of DNG files.
51125 | C7B5 | DefaultUserCrop | Used in Raw IFD of DNG files.
51177 | C7E9 | DepthFormat | Used in IFD 0 of DNG files.
51178 | C7EA | DepthNear | Used in IFD 0 of DNG files.
51179 | C7EB | DepthFar | Used in IFD 0 of DNG files.
51180 | C7EC | DepthUnits | Used in IFD 0 of DNG files.
51181 | C7ED | DepthMeasureType | Used in IFD 0 of DNG files.
51182 | C7EE | EnhanceParams | Used in Enhanced IFD of DNG files.

Source: https://www.awaresystems.be/imaging/tiff/tifftags/private.html

**Reply by radarhere** [reply]
I'm not sure how to answer this. I can't easily help you track down... 132 EXIF tags.

If there is a specific piece of information you would like to get out of an image, you can mention the tag, attach the image, and we can investigate that.

We've discussed how to access the IFD0, Exif, GPS, Interop and IFD1 groups. IFD1 is a special case though, as it's not using a tag offset like the others, but is just the next IFD. There is also Makernote, but that's specific to the manufacturer, and Pillow only supports Fujifilm and Nintendo Makernote data at the moment. A large portion of the tags you listed mention DNG, which Pillow doesn't support at the moment, see #6482.

As for the general question of if it is possible to identify all IFD data... I think the following part of the [TIFF specification](https://www.itu.int/itudoc/itu-t/com16/tiff-fx/docs/tiff6.pdf) might explain things a bit further (emphasis mine).
> If you need more than 5 or 10 tags, Aldus suggests that you reserve a single private tag, define it as a LONG, and use its value as a pointer (offset) to a private IFD or other data structure of your choosing. Within that IFD, you can use whatever tags you want, since **no one else will know that it is an IFD unless you tell them.**

**Reply by NickAyesha** [reply]
Thank you so much for your response, your insight, and your offer to help with additional data.
In retrospect, I should not have posted the above list. It was only to demonstrate the other tags that are probably under the Private IFD, not to suggest I need all of those tags.

This is the ultimate list that I need right now, with an emphasis on the critical and highs.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/nicoletterichards/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/nicoletterichards/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


PRIORITY | TAG
-- | --
High | FileName = IMG_2115.jpg
Medium | FileSize = 3042285
Medium | FileType = JPEG
Medium | FileTypeExtension = JPG
Medium | MIMEType = image/jpeg
Medium | \| City = New York
Medium | \| State = NY
Medium | \| Country = USA
Critical | \| Headline = The DYNAMICSAXHELP Gala, New York - Thursday, October 22th,   2023
Critical | \| CountryCode = USA
High | \| Location = Cipriani
High | \| DocumentID =   adobe:docid:photoshop:c3cbf3d1-2d70-634a-ab77-645bf7f6b6e9
Critical | \| Creator = DYNAMICSAXHELP
Critical | \| Description = DYNAMICSAXHELP   dedicated to conque[snip]
Critical | \| Subject = Technology
Critical | \| \| ObjectName = Bradley Segundo, Martin Galespe, and Prince Luci Min   Burger-Ojos
Critical | \| \| Keywords = Technology

</body>

</html>

Looking forward to your response.

**Reply by radarhere** [reply]
Could you attach a copy of IMG_2115.jpg?

**Reply by NickAyesha** [reply]
Is it possible for me to send it to you in email?

**Comment by NickAyesha** [comment]
According to information I am finding via AI

It seems perhaps the data I am looking for is nestled under Sub-IFDs.
Wondering if it is possible to iterate through sub-IFDs -  I'll check that.

IFD Code (Hex) | IFD Code (Decimal) | IFD Name | Description
-- | -- | -- | --
0x0000 | 0 | Image | Contains basic image information.
0x8769 | **34665 | Exif | Stores metadata related to the image.**
0x8825 | 34853 | GPS | Contains geographic information.
0xA005 | 40965 | Interop | Holds interoperability data.
0x0201 | 513 | Thumbnail | An optional sub-IFD within Exif.
Custom | Custom | MakerNotes | Manufacturer-specific information.
Custom | Custom | Composite | May contain derived composite data.
Custom | Custom | **Sub-IFDs | Sub-IFDs within other IFDs**.

<br class="Apple-interchange-newline">

**Comment by NickAyesha** [comment]
@radarhere 
There are some high profile people and wanted to obscure their images
![The_Global_Lyme_Alliance_GLA_Gala_2023_NCR_2115](https://github.com/python-pillow/Pillow/assets/32434816/db11a2e4-1f0e-4a91-9c14-228073e29fbc)

**Reply by radarhere** [reply, marked as answer]
```python
import os
from PIL import Image, ExifTags, IptcImagePlugin
with Image.open("IMG_2115.jpg") as im:
    print("FileName =", im.filename)
    print("FileSize =", os.path.getsize("IMG_2115.jpg"))
    print("FileType =", im.format)
    print("FileTypeExtension =", os.path.splitext(im.filename)[1][1:].upper())
    print("MIMEType =", im.get_format_mimetype())

    description = im.getxmp()["xmpmeta"]["RDF"]["Description"]
    print("| City =", description["City"])
    print("| State =", description["State"])
    print("| Country =", description["Country"])
    print("| Headline =", description["Headline"])
    print("| CountryCode =", description["CountryCode"])
    print("| Location =", description["Location"])
    print("| DocumentID =", description["DocumentID"])
    print("| Creator =", description["creator"]["Seq"]["li"])
    print("| Description =", description["description"]["Alt"]["li"]["text"])
    print("| Subject =", description["subject"]["Bag"]["li"])
    print("|| ObjectName =", description["title"]["Alt"]["li"]["text"])
    print("|| Keywords =", IptcImagePlugin.getiptcinfo(im)[(2, 25)])
```
gives
```
FileName = IMG_2115.jpg
FileSize = 3042017
FileType = JPEG
FileTypeExtension = JPG
MIMEType = image/jpeg
| City = New York
| State = NY
| Country = USA
| Headline = The  redacted Gala, New York - Thursday, October 22th, 2023
| CountryCode = USA
| Location = Cipriani
| DocumentID = adobe:docid:photoshop:c3cbf3d1-2d70-634a-ab77-645bf7f6b6e9
| Creator = NICOLETTE C. RICHARDS
| Description = redacted, the leading nonprofit dedicated to conquering Lyme and other tick-borne diseases holds their annual redacted at Cipriani 42nd Street in New York City on Thursday, redacted
| Subject = ['Non profit', 'Cipriani', 'Ambassador Images', 'Fundraiser', 'Auction', 'redacted']
|| ObjectName = Bradley Segundo, Martin Galespe, and Prince Luci Min Burger-Ojos  Attend Theredacted Gala - Thursday, October redacted, 2023
|| Keywords = [b'Non profit', b'Cipriani', b'Ambassador Images', b'Fundraiser', b'Auction', b'redacted']
```

**Reply by NickAyesha** [reply]
Hi! I didn't get an email notification like usual, so I thought you were busy. Thanks so much for this code. I'll use this. You are awesome!!!

**Reply by NickAyesha** [reply]
Radarhere, thanks again.
Dropping this here for anyone encountering this error: XMP data cannot be read without defusedxml dependency

You need to import **defusedxml**
